### ISSUE TITLE
Improve RuntimeV2 Macro Error Detail

### ISSUE THREAD

**Issue by AurevoirXavier** [paritytech/polkadot-sdk#6997]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Request

```
#[frame_support::runtime]
mod runtime {
	#[runtime::runtime]
	#[runtime::derive(
		RuntimeCall,
		RuntimeEvent,
		RuntimeError,
		RuntimeOrigin,
		RuntimeFreezeReason,
		RuntimeHoldReason,
		RuntimeSlashReason,
		RuntimeLockId,
		RuntimeTask
	)]
	pub struct Runtime;

	#[runtime::pallet_index(0)]
	pub type System = wrong_pallet_name;
}
```

If I write something like this, the error log only complains about `Runtime`/`RuntimeCall`/`xxx` not being found, which makes debugging difficult.

We should prioritize emitting a `wrong_pallet_name` not found error at the top.

### Are you willing to help with this request?

Maybe (please elaborate above)

**Comment by bkchr** [comment]
> We should prioritize emitting a `wrong_pallet_name` not found error at the top.

The problem here is that we don't have any control over this in the macro. The macro can not find out that easily if `wrong_pallet_name` exists or not. Maybe with some code parsing and finding out if `wrong_pallet_name` is a name of a pallet in the `Cargo.toml` or an import to the file.

**Comment by AurevoirXavier** [comment]
> The problem here is that we don't have any control over this in the macro. The macro can not find out that easily if `wrong_pallet_name` exists or not. Maybe with some code parsing and finding out if `wrong_pallet_name` is a name of a pallet in the `Cargo.toml` or an import to the file.

Hmm, I haven’t looked into this in depth.

However, I must say that after switching to the v2 macro, the development experience has declined. Issues like this were much easier to identify in v1.

**Comment by bkchr** [comment]
> I’m familiar with Substrate, so when I encounter such errors, I can assume there’s likely an issue with the macro. However, other developers might not realize this. The errors only indicate `Runtime` not found, `Pallet` not found, or `Event` not found, which can be confusing.

Yeah for sure! I mean I don't wanted to say that the feedback is not valid. I agree with you, but we are also not generating these errors, they are coming from the compiler. But we can improve them, if we do what I said above.

**Comment by olegtyshcneko** [comment]
> > We should prioritize emitting a `wrong_pallet_name` not found error at the top.
> 
> The problem here is that we don't have any control over this in the macro. The macro can not find out that easily if `wrong_pallet_name` exists or not. Maybe with some code parsing and finding out if `wrong_pallet_name` is a name of a pallet in the `Cargo.toml` or an import to the file.

Checking Cargo.toml in the procedural macro, while possible, I would say is a not a really good practice, macroses in general should be treated as pure functions (given code -> output new one). But for the sake of usufulness it might be an option, though there still can be issues with parsingCargo.toml , for example it can be changed and changes might not be tracked correctly. There are other issues with parsing Cargo toml files. I am thinking - that employing build.rs to prepare metadata for the proceducal macro code might be an option to consider. I will try to take a look into it, this is really interesting task!
